### DISCUSSION TITLE
How to use populate when not directly passed to repository?

### DISCUSSION THREAD

**Discussion by zetti-caletti** [mikro-orm/mikro-orm#4475] [Q&A]
**Describe the bug**
If I want to build the FindOneOptions manually, all things like orderBy work, but populate don't. I am now wondering how I can build my populate options dynamically for example depending on a parameter of a service method "includeMedia: boolean" for example.

**To Reproduce**
```
let findOneOptions: FindOneOptions<Post> = {
    orderBy: {
        media: {
            orderIndex: 'ASC',
            file: {
                name: 'ASC'
            }
        }
    },
    // Type 'string' is not assignable to type 'never'
    populate: ['media']
};
return this.postRepo.findOne({ id }, findOneOptions);
```

**Expected behavior**
Should work

**Additional context**

**Versions**

| Dependency | Version |
| - | - |
| node | ? |
| typescript | ? |
| mikro-orm | 5.6.15 |
| your-driver | ? |

**Comment by B4nan** [comment]
You are typing the object to `FindOneOptions<Post>` which means `FindOneOptions<Post, never>`, which means the populate hint is empty. TS does not allow partial type inference (although there are some improvements scheduled for 5.2, let's hope https://github.com/microsoft/TypeScript/pull/26349 will be merged).

In your particular example, you can do either `FindOneOptions<Post, any>`, which will allow you to add anything to the populate hint, or `FindOneOptions<Post, 'media'>` to have it strictly typed as if only the `media` relation is populated.

The real question is what you are trying to do, as this snippet is not very usable in general. If you want to do this on some abstract level, you need to use generics, just like the ORM does (check the definition of the EM methods).

**Reply by nicholasgubbins** [reply]
can you say a bit more about checking the definition of the EM methods. I'm facing a similar issue, as I've built a custom entity repository so that I can wrap each entity manager in a customer scoped transaction. See below

I've registered this customentity repository in my MikroOrm init - and now as I try to use it, I'm getting the same error as the previous commenter. I'm not sure how I should extend my custom repository to adopt the appropriate hints




```
import { EntityManager, EntityRepository } from "@mikro-orm/postgresql";

export class CustomEntityRepository<
  T extends object,
> extends EntityRepository<T> {

 async findOneForTenant(
    tenant_id: string,
    where: FilterQuery<T>,
    options?: FindOneOptions<T>,
  ): Promise<T | null> {
    return await this.getEntityManager().transactional(async (em) => {
      await em.execute(
        'SELECT set_config('app.current_tenant',?::text,true);',
        [tenant_id],
      );
      return em.findOne(this.entityName, where, options);
    });
  }

  async findOneOrFailForTenant(
    tenant_id: string,
    where: FilterQuery<T>,
    options?: FindOneOrFailOptions<T>,
  ): Promise<T> {
    return await this.getEntityManager().transactional(async (em) => {
      await em.execute(
        'SELECT set_config('app.current_tenant',?::text,true);',
        [tenant_id],
      );
      return em.findOneOrFail(this.entityName, where, options);
    });
  }
}

```

**Reply by nicholasgubbins** [reply]
on the above the error is being thrown here: 

```
import { IQueryHandler, QueryHandler } from "@nestjs/cqrs";
import { InjectRepository } from "@mikro-orm/nestjs";
import { CustomEntityRepository as EntityRepository } from "../../database/custom-entity-repository";
import { GetTalentTypeByIdQuery } from "../queries/get-talent-type-by-id.query";
import TalentTypeEntity from "../entities/talent-type.entity";

@QueryHandler(GetTalentTypeByIdQuery)
export class GetTalentTypeByIdHandler
  implements IQueryHandler<GetTalentTypeByIdQuery>
{
  constructor(
    @InjectRepository(TalentTypeEntity)
    private talenttypeRepository: EntityRepository<TalentTypeEntity>,
  ) {}

  async execute(query: GetTalentTypeByIdQuery): Promise<TalentTypeEntity> {
    //find talenttypes where
    const talenttype = await this.talenttypeRepository.findOneForTenant(
      query.tenant_id,
      {
        id: query.talenttype_id,
      },
      {
        populate: ["custom_property_settings"],
      },
    );
    return talenttype;
  }
}
```

![image](https://github.com/mikro-orm/mikro-orm/assets/6475716/a207602a-6c3c-424f-b51d-eec80bf034dc)

**Reply by B4nan** [reply]
You need to define the populate hint as a type argument, something like this:

```ts
 async findOneForTenant<Hint extends string = never>(
    tenant_id: string,
    where: FilterQuery<T>,
    options?: FindOneOptions<T, Hint>,
  ): Promise<T | null> {
    return await this.getEntityManager().transactional(async (em) => {
      await em.execute(
        'SELECT set_config('app.current_tenant',?::text,true);',
        [tenant_id],
      );
      return em.findOne(this.entityName, where, options);
    });
  }
```

In v6 there is also the partial loading hint type argument, so there it would be like this:

```ts
 async findOneForTenant<Hint extends string = never, Fields extends string = never>(
    tenant_id: string,
    where: FilterQuery<T>,
    options?: FindOneOptions<T, Hint, Fields>,
  ): Promise<T | null> {
    return await this.getEntityManager().transactional(async (em) => {
      await em.execute(
        'SELECT set_config('app.current_tenant',?::text,true);',
        [tenant_id],
      );
      return em.findOne(this.entityName, where, options);
    });
  }
```

**Reply by nicholasgubbins** [reply]
makes sense, much apprciated
